MapboxNavigation/NavigationMapView.swift

```diff
@@ -54,56 +51,248 @@ open class NavigationMapView: MGLMapView {
+        UIView.animate(withDuration: duration, delay: 0, options: [.curveLinear, .beginFromCurrentState], animations: {
+            self.userCourseView?.center = self.convert(location.coordinate, toPointTo: self)
+        }, completion: nil)
+        
+        if let userCourseView = userCourseView as? UserCourseView {
+            userCourseView.update(location: location, pitch: camera.pitch, direction: direction, animated: animated, tracksUserCourse: tracksUserCourse)
         }
     }
     
-    /**
-     Force MGLMapView to ask its delegate for the user location annotation’s view, in case that previously happened before there was a delegate.
-     */
-    func refreshShowsUserLocation() {
-        showsUserLocation = false
-        showsUserLocation = true
+    var targetPoint: CGPoint {
+        return CGPoint(x: bounds.midX, y: bounds.height*0.75)
     }
     
-    override open func locationManager(_ manager: CLLocationManager!, didUpdateLocations locations: [CLLocation]!) {
-        guard let location = locations.first else { return }
+    var tracksUserCourse: Bool = false {
+        didSet {
+            if tracksUserCourse {
+                showsUserLocation = true
+                courseTrackingDelegate?.navigationMapViewDidStartTrackingCourse(self)
+            } else {
+                courseTrackingDelegate?.navigationMapViewDidStopTrackingCourse(self)
+            }
+            
+            if let location = userLocationForCourseTracking {
+                updateCourseTracking(location: location, animated: true)
+            }
+        }
+    }
+    
+    open override func mapViewDidFinishRenderingFrameFullyRendered(_ fullyRendered: Bool) {
+        super.mapViewDidFinishRenderingFrameFullyRendered(fullyRendered)
         
-        if let modifiedLocation = navigationMapDelegate?.navigationMapView?(self, shouldUpdateTo: location) {
-            super.locationManager(manager, didUpdateLocations: [modifiedLocation])
-        } else {
-            super.locationManager(manager, didUpdateLocations: locations)
+        guard shouldPositionCourseViewFrameByFrame else { return }
+        guard let location = userLocationForCourseTracking else { return }
+        
+        userCourseView?.center = convert(location.coordinate, toPointTo: self)
+    }
+    
+    /**
+     A `UIView` used to indicate the user’s location and course on the map.
+     
+     If the view conforms to `UserCourseView`, its `UserCourseView.update(location:pitch:direction:animated:)` method is frequently called to ensure that its visual appearance matches the map’s camera.
+     */
+    public var userCourseView: UIView? {
+        didSet {
+            if let userCourseView = userCourseView {
+                addSubview(userCourseView)
+            }
         }
     }
     
-    override open func validateLocationServices() {
-        if !manuallyUpdatesLocation {
-            super.validateLocationServices()
+    var userCourseViewCenter: CGPoint {
```

Comment:
This variable is unused, because `targetPoint` was added at some point to do the same thing. `targetPoint` is pretty unsophisticated; can we go back to something like this that respects the content insets?